src/jdk.jfr/share/classes/jdk/jfr/internal/StringPool.java

```diff
@@ -71,27 +73,49 @@ private static long externalSid(long internalSid) {
         return internalSid >> 16;
     }
 
-    /* synchronized because of writing the string to the JVM. */
-    private static synchronized long storeString(String s) {
-        Long lsid = cache.get(s);
-        long internalSid;
-        if (lsid != null) {
-            internalSid = lsid.longValue();
-            if (isCurrentGeneration(internalSid)) {
-                // Someone already updated the cache.
-                return externalSid(internalSid);
+    /* Explicitly pin a virtual thread before acquiring the string pool monitor
+     * because migrating the EventWriter onto another carrier thread is impossible.
+     */
+    private static void pinVirtualThread() {
+        if (Thread.currentThread().isVirtual() && ContinuationSupport.isSupported()) {
+            Continuation.pin();
+        }
+    }
+
+    private static void unpinVirtualThread() {
+        if (Thread.currentThread().isVirtual() && ContinuationSupport.isSupported()) {
```

Comment:
If you are at all concerned about overhead here then pin could return a boolean to indicate if the pin happened and oyu could then unpin just by checking that boolean and avoid doing the isVirtual and isSupported checks again.

Reply:
Would be possible to create a boolean in the EventWriter that indicates if it is associated with a carrier thread or a normal thread (which can never be virtual) and then have two methods.

    long l = carrier ? StringPool.addPinnedString(s) : StringPool.addString(s);

Reply:
Thread.currentThread() has an intrinsic, and isVirtual is just a type check. ContinuationSupport.isSupported reads a static final so will disappear once compiled. The pattern we are using in other areas is for the pin to return a boolean (like David suggested).

Reply:
I looked into this in more detail. The current suggestion:

mov    r10,QWORD PTR [r15+0x388]  ; _vthread OopHandle
mov    r10,QWORD PTR [r10]              ; dereference OooHandle <<-- Thread.currentThread() intrinsic gives 2 instructions
mov    r11d,DWORD PTR [r10+0x8]    ; InstanceKlass to r11
mov    r10d,r11d                                 ; InstanceKlass to r10
mov    r8,QWORD PTR [r10+0x40]      ; Load slot in InstanceKlass primary supers array to r8
movabs r10,0x2d0481a8                     ; InstanceKlass for {metadata('java/lang/BaseVirtualThread')} to r10
cmp    r8,r10                                       ; compare if superklass is java/lang/BaseVirtualThread
jne    0x0000018571e0baf9                ; 6 instructions for isVirtual() type check, 8 instructions in total

This gives a prologue for eight instructions. We already have much of this information resolved when loading up the EventWriter instance using the existing intrinsic getEventWriter(). Hence, we could extend that to mark the event writer with a field to say if pinning should be performed. This results in only a two instruction prologue:

test   r8d,r8d                         ; pinVirtualThread? 
je     0x0000012580a0f6c9    ; 2 instructions for test

This is an x4 speedup, although slightly less because of an additional store instruction for loading the event writer.

Further, I looked into the actual Continuation.pin() and Continuation.unpin() instructions. They are currently not intrinsics, but they lend themselves well to intrinsification. I have created such intrinsics, and the results are quite good.

Continuation.pin() or Continuation.unpin() without intrinsics = 112 instructions each
Continuation.pin() or Continuation.unpin() with intrinsics = 8 instructions each

This is an x14 speedup for virtual threads.

Reply:
I plan to fix the event writer under this PR (to be updated) and file a separate tracking enhancement for the intrinsification of the pin and unpin.